File: src/server/services/memory/memory.go
package memory

import (
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

// MemoryService interface for agent episodic memory cache
type MemoryService interface {
	SaveEpisodicMemory(ctx context.Context, tenantID, sessionID, context string) error
	RetrieveRecentMemory(ctx context.Context, tenantID, sessionID string) (string, error)
}

// AgentMemoryService implementation
type AgentMemoryService struct {
	client *redis.Client
}

// NewAgentMemoryService creates a new memory service
func NewAgentMemoryService(client *redis.Client) *AgentMemoryService {
	return &AgentMemoryService{
		client: client,
	}
}

// key constructs the isolated tenant key
func key(tenantID, sessionID string) string {
	return fmt.Sprintf("ohc:mem:%s:%s", tenantID, sessionID)
}

// SaveEpisodicMemory saves memory to Redis cache strictly isolated by tenant ID
func (s *AgentMemoryService) SaveEpisodicMemory(ctx context.Context, tenantID, sessionID, memoryContext string) error {
	return s.client.Set(ctx, key(tenantID, sessionID), memoryContext, 7*24*time.Hour).Err()
}

// RetrieveRecentMemory gets recent memory isolated by tenant ID
func (s *AgentMemoryService) RetrieveRecentMemory(ctx context.Context, tenantID, sessionID string) (string, error) {
	val, err := s.client.Get(ctx, key(tenantID, sessionID)).Result()
	if err == redis.Nil {
		return "", nil // Return empty string for not found instead of error
	} else if err != nil {
		return "", err
	}
	return val, nil
}

File: src/server/services/memory/src/memory.rs
use redis::{Client, Commands, RedisResult};
use std::time::Duration;

pub struct AgentMemoryService {
    client: Client,
}

impl AgentMemoryService {
    pub fn new(client: Client) -> Self {
        AgentMemoryService { client }
    }

    fn key(tenant_id: &str, session_id: &str) -> String {
        format!("ohc:mem:{}:{}", tenant_id, session_id)
    }

    pub fn save_episodic_memory(
        &self,
        tenant_id: &str,
        session_id: &str,
        context: &str,
    ) -> RedisResult<()> {
        let mut con = self.client.get_connection()?;
        let key = Self::key(tenant_id, session_id);

        let _: () = redis::cmd("SETEX")
            .arg(key)
            .arg(7 * 24 * 60 * 60) // 7 days in seconds
            .arg(context)
            .query(&mut con)?;

        Ok(())
    }

    pub fn retrieve_recent_memory(
        &self,
        tenant_id: &str,
        session_id: &str,
    ) -> RedisResult<Option<String>> {
        let mut con = self.client.get_connection()?;
        let key = Self::key(tenant_id, session_id);

        let val: Option<String> = con.get(key)?;
        Ok(val)
    }
}
